tests/server/tso/allocator_test.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tso_test
+
+import (
+	"context"
+	"path"
+	"time"
+
+	. "github.com/pingcap/check"
+	"github.com/tikv/pd/pkg/slice"
+	"github.com/tikv/pd/server"
+	"github.com/tikv/pd/server/election"
+	"github.com/tikv/pd/server/tso"
+	"github.com/tikv/pd/tests"
+)
+
+const waitAllocatorCheckInterval = 500 * time.Millisecond
+
+var _ = Suite(&testAllocatorSuite{})
+
+type testAllocatorSuite struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (s *testAllocatorSuite) SetUpSuite(c *C) {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	server.EnableZap = true
+}
+
+func (s *testAllocatorSuite) TearDownSuite(c *C) {
+	s.cancel()
+}
+
+// Make sure we have the correct number of allocator leaders.
+func (s *testAllocatorSuite) TestAllocatorLeader(c *C) {
+	var err error
+	cluster, err := tests.NewTestCluster(s.ctx, 3)
+	defer cluster.Destroy()
+	c.Assert(err, IsNil)
+
+	err = cluster.RunInitialServers()
+	c.Assert(err, IsNil)
+
+	ctx, cancel := context.WithCancel(s.ctx)
+	defer cancel()
+	// There will be the number of dc-locations Local TSO Allocator leaders elected
+	testDCLocations := []string{"dc-1", "dc-2", "dc-3"}
+	dcLocationNum := len(testDCLocations)
+	for _, dcLocation := range testDCLocations {
+		for _, server := range cluster.GetServers() {
+			tsoAllocatorManager := server.GetTSOAllocatorManager()
+			leadership := election.NewLeadership(
+				server.GetEtcdClient(),
+				path.Join(server.GetServer().GetServerRootPath(), dcLocation),
+				"campaign-local-allocator-test")
+			tsoAllocatorManager.SetUpAllocator(ctx, cancel, dcLocation, leadership)
+
+		}
+	}
+	// To check whether we have enough Local TSO Allocator leaders
+	allAllocatorLeaders := make([]tso.Allocator, 0, dcLocationNum)
+	for i := 0; i < 20; i++ {
```

Comment:
it tooks 10s, can the test time be reduced? such as break if already have two local TSO allocator leader.

Reply:
I do this on purpose. If we break the loop when we have two local TSO allocator leaders as soon as possible, then the test will be meaningless. We need to forcibly wait for a while to make sure after this duration, only two local TSO allocator leaders are elected.

Reply:
yes, but it's may re-election a new etcd leader in the unstable environment of CI, it's may appear three local allocators in history, it is a normal situation but the test will be failed.

Reply:
and my point is that we should take the test as short as possible

Reply:
I add the member number to 6. And I change the way to wait that we only wait for 1 second before checking. It seems that is long enough for this test. Nice suggest!